feat(presence): Raise events for presence notifications

tylerbutler opened the pull request:
Events for presence notifications have been implemented. Some type-related workarounds were needed, but all are documented inline.

Two tests have been added:

1. Validates that sending notifications via the emit.broadcast function causes the correct signal to be sent.
2. Validates that event handler functions are called when the notification signals are received.

Note that the emit.unicast function, used to send notifications to specific clients, is not tested.

[AB#19706](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/19706)

tylerbutler committed:
wip

tylerbutler committed:
wip

tylerbutler committed:
wip

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
compiles and test passes

tylerbutler committed:
process incoming signal

tylerbutler committed:
sort of works

tylerbutler committed:
cleanup

tylerbutler committed:
updates

tylerbutler commented on packages/framework/presence/src/test/tsconfig.json:
```diff
@@ -7,7 +7,7 @@
 		"types": ["mocha", "node"],
 		"noImplicitAny": true,
-		"noUnusedLocals": true,
+		"noUnusedLocals": false,
```
disabled for easier testing

tylerbutler replied:
```suggestion
		"noUnusedLocals": true,
```

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -152,6 +152,7 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
 		): void => {
 			// Check for connectivity before sending updates.
 			if (!this.runtime.connected) {
+				console.debug(`this.runtime is not connected`);
```
```suggestion
```

tylerbutler commented on packages/framework/presence/src/test/core-interfaces/testValues.ts:
```diff
@@ -349,29 +349,29 @@ export const arrayOfLiterals: readonly (
 export class ClassWithPrivateData {
 	public public = "public";
-	// @ts-expect-error secret is never read
+	// // @ts-expect-error secret is never read
```
```suggestion
	// @ts-expect-error secret is never read
```

tylerbutler commented on packages/framework/presence/src/test/core-interfaces/testValues.ts:
```diff
@@ -349,29 +349,29 @@ export const arrayOfLiterals: readonly (
 export class ClassWithPrivateMethod {
 	public public = "public";
-	// @ts-expect-error getSecret is never read
+	// // @ts-expect-error secret is never read
```
```suggestion
	// @ts-expect-error secret is never read
```

tylerbutler commented on packages/framework/presence/src/test/core-interfaces/testValues.ts:
```diff
@@ -349,29 +349,29 @@ export const arrayOfLiterals: readonly (
 export class ClassWithPrivateGetter {
 	public public = "public";
-	// @ts-expect-error secret is never read
+	// // @ts-expect-error secret is never read
```
```suggestion
	// @ts-expect-error secret is never read
```

tylerbutler commented on packages/framework/presence/src/test/core-interfaces/testValues.ts:
```diff
@@ -349,29 +349,29 @@ export const arrayOfLiterals: readonly (
 export class ClassWithPrivateSetter {
 	public public = "public";
-	// @ts-expect-error secret is never read
+	// // @ts-expect-error secret is never read
```
```suggestion
	// @ts-expect-error secret is never read
```

tylerbutler commented on packages/framework/presence/src/test/mockEphemeralRuntime.ts:
```diff
@@ -135,6 +135,7 @@ export class MockEphemeralRuntime implements IEphemeralRuntime {
 			throw new Error(`Unexpected signal: ${JSON.stringify(args)}`);
 		}
 		const expected = this.signalsExpected.shift();
+		// console.debug(JSON.stringify(args, undefined, 2));
```
```suggestion
```

tylerbutler committed:
Apply suggestions from code review

tylerbutler commented on packages/framework/presence/src/test/core-interfaces/testValues.ts:
```diff
@@ -355,7 +355,7 @@ export class ClassWithPrivateData {
 export class ClassWithPrivateMethod {
 	public public = "public";
-	// @ts-expect-error getSecret is never read
+	// @ts-expect-error secret is never read
```
```suggestion
	// @ts-expect-error getSecret is never read
```

tylerbutler committed:
Apply suggestions from code review

tylerbutler committed:
revert changes in example

tylerbutler committed:
cleanup

tylerbutler commented on packages/framework/presence/src/test/notificationsManager.spec.ts:
```diff
@@ -3,16 +3,268 @@
+			logger = new EventAndErrorTrackingLogger();
+			runtime = new MockEphemeralRuntime(logger);
+
+			// TODO: runtime.connected has been hacked in past to lie about true connection.
+			// This will need to be updated to an alternate status provider.
+			// runtime.connected = true;
```
This hasn't been necessary in my testing.

jason-ha replied:
Update: You don't need this because `prepareConnectedPresence` does it.

jason-ha commented on packages/framework/presence/src/test/notificationsManager.spec.ts:
```diff
@@ -3,16 +3,268 @@
+			// Set up the presence connection, which will create an expected signal, so add that signal to the expected list
+			// as well
+			presence = prepareConnectedPresence(runtime, "sessionId-2", "client2", clock, logger);
+			runtime.signalsExpected.push(
+				// The first signal, the join signal, is from the prepareConnectedPresence call
+				// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+				knownSignals.get("PresenceJoin")!,
+			);
```
This doesn't add up for me.
1) `prepareConnectedPresence` already takes care of pushing the signal expected from it (see https://github.com/tylerbutler/FluidFramework/blob/presence-notification-events/packages/framework/presence/src/test/testUtils.ts#L96)
2) signalsExpected should always be added before they might happen (the array is shortened as events happen)

I could see that `prepareConnectedPresence` doesn't submit signal here because `runtime.connected` was not set to `true`. Something though must convince PresenceManager that it is connected at some point for signals to get submitted. I don't see later code explicit changing that state.
Hmm...

jason-ha replied:
I debugged. What is happening is that there is a Join signal process just below. That Join requests that the local client ("client2") respond. The signal expected is the one sent to satisfy that request. It doesn't look like anything relies on there being another client present. So, could just remove the expected signal push and the processSignal below.

tylerbutler replied:
Thanks - I cleaned things up quite a bit in the latest version. Working on the core code changes now.

jason-ha reviewed (commented):
posting partial review notes - something I don't understand with the test setup currently

jason-ha commented on packages/framework/presence/src/test/notificationsManager.spec.ts:
```diff
@@ -3,16 +3,268 @@
+				"isComplete": true,
+				"data": {
+					"system:presence": {
+						"clientToSessionId": {
+							"client2": { "rev": 0, "timestamp": 1000, "value": "sessionId-2" },
+						},
+					},
+				},
+			},
+		],
+	],
+
+	[
+		"setNewId42",
+		[
+			"Pres:DatastoreUpdate",
+			{
+				"sendTimestamp": 1020,
+				"avgLatency": 10,
+				"data": {
+					"system:presence": {
+						"clientToSessionId": {
+							"client2": { "rev": 0, "timestamp": 1000, "value": "sessionId-2" },
+						},
+					},
+					"n:name:testNotificationWorkspace": {
+						"testEvents": {
+							"sessionId-2": {
+								"rev": 0,
+								"timestamp": 0,
+								"value": { "name": "newId", "args": [42] },
+								"ignoreUnmonitored": true,
+							},
+						},
+					},
+				},
+			},
+		],
+	],
+]);
```
Shouldn't need to put this in a Map and deal with `get` that might fail. I think can just use a record. (Maybe your snapshot tool is more compatible with this, but we can update it)
```suggestion
/**
 * A record of presence signals used for testing, indexed by a friendly string.
 */
const knownSignals = {
	"PresenceJoin": [
		"Pres:DatastoreUpdate",
		{
			"sendTimestamp": 1010,
			"avgLatency": 10,
			"isComplete": true,
			"data": {
				"system:presence": {
					"clientToSessionId": {
						"client2": { "rev": 0, "timestamp": 1000, "value": "sessionId-2" },
					},
				},
			},
		},
	],
	"setNewId42": [
		"Pres:DatastoreUpdate",
		{
			"sendTimestamp": 1020,
			"avgLatency": 10,
			"data": {
				"system:presence": {
					"clientToSessionId": {
						"client2": { "rev": 0, "timestamp": 1000, "value": "sessionId-2" },
					},
				},
				"n:name:testNotificationWorkspace": {
					"testEvents": {
						"sessionId-2": {
							"rev": 0,
							"timestamp": 0,
							"value": { "name": "newId", "args": [42] },
							"ignoreUnmonitored": true,
						},
					},
				},
			},
		},
	],
} as const satisfies Record<string, Parameters<IEphemeralRuntime["submitSignal"]>>;
```

Then you can just use `knownSignals.<name>` where needed.

MockEphemeralRuntime ctor will need a parameter update. It is missing readonly qualification:
```diff
-		public readonly signalsExpected: Parameters<IEphemeralRuntime["submitSignal"]>[] = [],
+		public readonly signalsExpected: Readonly<
			Parameters<IEphemeralRuntime["submitSignal"]>
		>[] = [],
```

tylerbutler replied:
ended up ditching the map completely since it wasn't saving any typing and made the individual tests less clear. Can bring it back if there are more tests. Or refactor the testUtils stuff to make the signals a bit more reusable.

tylerbutler committed:
simplify and correct tests

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
feedback

tylerbutler committed:
build

tylerbutler committed:
Merge branch 'main' into presence-notification-events

jason-ha committed:
test(client-presence): rework Notifications tests
- move workspace create to common setup
- add more test cases
- rework emit test without spies
- remove console logging
- remove compile coverage as part of regular tests

jason-ha reviewed (commented):
I thought the tests could be improved and made this PR to your branch: https://github.com/tylerbutler/FluidFramework/pull/30

Two of the tests are failing. Looks like that is because initial subscriptions are not being added as listeners. (Probably what led you to add the extra listeners.)

tylerbutler committed:
Merge pull request #30 from microsoft/presence/notification-events-tests

test(client-presence): rework Notifications tests

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
Merge branch 'main' into presence-notification-events

tylerbutler committed:
fix initial subscriptions not being triggered

tylerbutler commented on packages/framework/presence/src/notificationsManager.ts:
```diff
@@ -170,8 +170,16 @@ class NotificationsManagerImpl<
+	) {
+		// Add event listeners provided at instantiation
+		for (const [nameString, value] of Object.entries(initialSubscriptions)) {
+			// Without schema validation, we don't know that the args are the correct type.
+			// For now we assume the user is sending the correct types and there is no corruption along the way.
```
I don't think this comment is correct.

tylerbutler replied:
The casts need necessary, but are there better types to use?

jason-ha replied:
Use this for the loop:
```typescript
		for (const subscriptionName of recordKeys(initialSubscriptions)) {
			// Lingering Event typing issues with Notifications specialization requires
			// this cast. The ony thing that really matters is that name is a string.
			const name = subscriptionName as keyof Events<NotificationSubscriptions<T>>;
			const value = initialSubscriptions[subscriptionName];
			// This check should not be needed while using exactOptionalPropertyTypes, but
			// typescript appears to ignore that with Partial<>. Good to be defensive
			// against callers sending `undefined` anyway.
			if (value !== undefined) {
				this.notificationsInternal.on(name, value);
			}
```
`Object.entries` loses value typing; so, enumerating keys and getting the value will preserve that aspect.
`Object.key` loses the key typing as well, but we can adjust for that with:
```typescript
/**
 * Object.keys retyped to support specific records keys and
 * branded string-based keys.
 */
const recordKeys = Object.keys as <K extends string>(o: Partial<Record<K, unknown>>) => K[];
```
(add locally above the class)
That isn't quite enough - hence cast and comment that maybe can go away some day.

tylerbutler committed:
Remove outdated comments

tylerbutler committed:
Merge branch 'main' into presence-notification-events

jason-ha commented on packages/framework/presence/src/test/notificationsManager.spec.ts:
```diff
@@ -264,9 +264,12 @@ describe("Presence", () => {
-			assert(eventHandlerCalls.original.length === 1);
-			assert(eventHandlerCalls.secondary.length === 1);
-			assert(eventHandlerCalls.tertiary.length === 1);
+			assert(eventHandlerCalls.original.length === 1, "original event handler was not called");
```
I like a helpful message. It could be that event handler was called more than once.

tylerbutler replied:
I included the actual value in the message so that should clarify.

jason-ha commented on packages/framework/presence/src/test/notificationsManager.spec.ts:
```diff
@@ -3,86 +3,480 @@
+					},
+					"testEvents"
+				>({
+					newId: (_client: ISessionClient, _id: number) => {},
```
Let's also test that an initial subscription is not required by omitting one here.
It looks like `Notifications` function will need to update argument to be partial:
```
	initialSubscriptions: Partial<NotificationSubscriptions<T>>,
```

tylerbutler committed:
clean up typing

tylerbutler committed:
Merge branch 'main' into presence-notification-events

jason-ha commented on packages/framework/presence/src/notificationsManager.ts:
```diff
@@ -173,14 +179,19 @@ class NotificationsManagerImpl<
+			}
 		}
 	}
-
```
nit: dropped vertical spacing (we can fix later)

tylerbutler committed:
feedback

jason-ha commented on packages/framework/presence/src/test/notificationsManager.spec.ts:
```diff
@@ -264,12 +265,18 @@ describe("Presence", () => {
+				eventHandlerCalls.tertiary.length === 1,
+				`secondary event handler was called ${eventHandlerCalls.tertiary.length} times; expected 1`,
 			);
-			assert(eventHandlerCalls.tertiary.length === 1, "tertiary event handler was not called");
```
What happened to this third guy?

jason-ha commented on packages/framework/presence/src/test/notificationsManager.spec.ts:
```diff
@@ -264,12 +265,18 @@ describe("Presence", () => {
 			);
 
-			assert(eventHandlerCalls.original.length === 1, "original event handler was not called");
+			assert(
+				eventHandlerCalls.original.length === 1,
+				`original event handler was called ${eventHandlerCalls.original.length} times; expected 1`,
+			);
```
`assert.equal(eventHandlerCalls.original.length, 1` is preferred (easier for authors)

tylerbutler committed:
formatting
